c++/src/kj/async-io.c++

```diff
@@ -2281,6 +2317,20 @@ public:
     return tee->tryGetLength(branch);
   }
 
+  Maybe<Own<AsyncInputStream>> tryTee(uint64_t limit) override {
+    if (tee->getBufferSizeLimit() > limit) {
+      // Cannot optimize this path as the limit has shrunk further. A new explicit tee branch is
+      // required to handle this case. Growth should be fine because it means someone tried to do
+      // something like:
+      //    auto [b, c] = newTee(a, 5)
+      //    auto [d, e] = newTee(b, 10)
+      // The limit of 10 can be ignored because the root stream has a limit of 5.
```

Comment:
I *think* I may have misunderstood how the limit works. I thought that the limit applies to each individual branch.
```
      for (auto& state: branches) {
        KJ_IF_MAYBE(s, state) {
          // TODO(perf): buffer.size() is O(n) where n = # of individual heap-allocated byte arrays.
          if (s->buffer.size() + n.maxBytes > bufferSizeLimit) {
            stoppage = Stoppage(KJ_EXCEPTION(FAILED, "tee buffer size limit exceeded"));
            return pullLoop();
          }
        }
      }
```

So in your example, I'm not sure why `b` being destroyed changes any behavior but I *think* the implication must be that the sum of all buffers used by all branches must be within that memory limit? Is there somewhere else that's happening? If not, this may be an instance of a possible memory explosion that could be constructed since each tee has an independent buffer of the underlying stream without that being exposed to V8, allowing an accidental memory explosion to exceed the specified budget...?

Similarly:
>   // I need to be able to buffer at least a megabyte, but I know I won't go over two megabytes.
>  auto [d, e] = newTee(stream, 2 * (1 << 20));
>  // At this point, I think `d` and `e` can buffer at least a megabyte, but that's not true.

My understanding of the current system is that `newTee` would create an `AsyncTee` wrapping `c` & applying a 2MB limit. However, `c` itself would be a branch pointing to an `AsyncTee` that's limited to 128 so it would end up not really being able to buffer any more than 128 anyway. Unless you're saying that today using a new limit of 2MB would result in reading the underlying stream in 128 byte chunks but you would still be allowed to buffer 2MB?

Aside:
This seems weird:
```
    // If there is excess data in the buffer for us, slurp that up.
    auto readBuffer = arrayPtr(reinterpret_cast<byte*>(buffer), maxBytes);
    auto readSoFar = state.buffer.consume(readBuffer, minBytes);

    if (minBytes == 0) {
      return readSoFar;
    }
```
Shouldn't that be `if (minBytes <= readSoFar && readSoFar)`? Otherwise if I ask for min 0 bytes & there's nothing in the buffer, I'll never read anything...?

Reply:
Discussed in person everything before the aside. Seems reasonable that this is a very coarse protection mechanism & not actually useful to worry about the absolute memory used.

@harrishancock would be good to get clarification on the aside in case there's a latent issue here that might be need to be fixed?

Reply:
Gah - nvm. I missed that `state.buffer.consume` takes `minBytes` by reference. Looks all good here.

Reply:
I think we got on the same page yesterday, but just for clarity:

> I think the implication must be that the sum of all buffers used by all branches must be within that memory limit?

As implemented, the limit is applied on each branch's buffers individually. That is, if the limit is 10, then branch _a_ could buffer up to 10 bytes, and branch _b_ could also buffer an additional 10 bytes. This has not been a practical concern until now because the only time we buffer data is when one of the branches initiates a read operation, in which case we immediately flush one of the branch's buffers. By the time we enforce the limit, only one branch could have buffers in it.

That will change when supporting n-tees. Considering the case of a 3-tee, a read operation by one branch would immediately be flushed, but the other two branches would each end up buffering a separate copy of the same data, meaning the maximum amount of data we might end up buffering is actually `(n - 1) * limit`, where _n_ is the number of tee branches.

For now, I think we shouldn't change this -- the new behavior is no worse than the current situation, and if we were to enforce the limit on all buffers combined, that would be a potentially breaking behavior change.

In the future, we could switch to a shared buffer, and have each branch keep owning views onto the shared buffer. E.g., if branch 0 hasn't read anything, branch 1 has read 5 bytes, and branch 2 has read 10 bytes, we could store only 10 bytes, not 15, and keep per-branch indices into the buffer, where branch 0 would start at index 0, branch 1 would start at index 5, and branch 3 would start at index 10. When branch 0 (the laggard branch) reads a byte, we have an opportunity to resize the buffer smaller, if we wished. I initially tried to implement that when I wrote this, but got confused and stressed out and ended up with this naive implementation.

There's also a philosophy where we might pass in some buffer implementation, giving the user of `newTee()` more control, but I'm not sure what the interface would be like.

> Unless you're saying that today using a new limit of 2MB would result in reading the underlying stream in 128 byte chunks but you would still be allowed to buffer 2MB? But in that case I don't know how to implement the new limit in a way that makes sense - seems like there's a lot of nuance & some corner condition will always not work?

Yes, today using a new limit of 2MB would have that effect: the underlying stream would be read in at most 128 byte chunks due to the nested tee (which seems like sort of a bug -- a branch of a 2-tee whose counterpart is destroyed should ideally revert to whatever behavior the stream had before it was teed, I think, but oh well), and the new wrapping tee would allow up to 2MB of data buffered.

If a 2-tee's branches both still exist, and we try to apply another tee with a different limit to one of the branches, then I think we must resort to nesting in order to continue to allow users to specify high limits and being able to rely on reading at least that many bytes from one of the branches.

If a tee has become a 1-tee -- i.e., only one surviving branch -- and we try to apply a different limit, then it might make sense to _replace_ the existing AsyncTee's limit with the new limit. Note that a 1-tee never buffers (or does so only transiently during the course of an I/O operation), so the buffer limit should be meaningless to it. (If the new limit were less than the old, and the surviving branch is behind its old sibling in the stream, then there could be situations where we would temporarily be exceeding the new limit until the branches drained their buffers sufficiently. I don't think that's a problem, though.)

All that said, it's probably best if we just unconditionally resort to nesting if the new limit is different for now -- it should always be correct.